src/runtime_src/core/edge/user/device_linux.cpp

```diff
@@ -159,6 +174,90 @@ struct kds_cu_info
   }
 };
 
+struct aie_reg_read
+{
+  using result_type = query::aie_reg_read::result_type;
+
+  static result_type
+  get(const xrt_core::device* device, key_type key, const boost::any& r, const boost::any& c, const boost::any& reg)
+  {
+    auto dev = get_edgedev(device);
+    uint32_t val = 0;
+    auto row = boost::any_cast<int>(r);
+    auto col = boost::any_cast<int>(c);
+    auto v = boost::any_cast<std::string>(reg);
+ 
+#ifdef XRT_ENABLE_AIE
+#ifndef __AIESIM__
+  const static std::string AIE_TAG = "aie_metadata";
+  const static std::string ZOCL_DEVICE = "/dev/dri/renderD128";
+
+  std::string err;
+  std::string value;
+  dev->sysfs_get(AIE_TAG, err, value);
+  if (!err.empty())
+    throw xrt_core::error(-EINVAL, err + ", aie xclbin not loaded");
+  std::stringstream ss(value);
+  boost::property_tree::ptree pt;
+  boost::property_tree::read_json(ss, pt);
+
+  int mKernelFD = open(ZOCL_DEVICE.c_str(), O_RDWR);
+  if (!mKernelFD)
+    throw xrt_core::error(-EINVAL, boost::str(boost::format("Cannot open %s") % ZOCL_DEVICE));
+
+  XAie_DevInst* devInst;         // AIE Device Instance
+
+  XAie_SetupConfig(ConfigPtr,
+    pt.get<uint8_t>("aie_metadata.driver_config.hw_gen"),
+    pt.get<uint64_t>("aie_metadata.driver_config.base_address"),
+    pt.get<uint8_t>("aie_metadata.driver_config.column_shift"),
+    pt.get<uint8_t>("aie_metadata.driver_config.row_shift"),
+    pt.get<uint8_t>("aie_metadata.driver_config.num_columns"),
+    pt.get<uint8_t>("aie_metadata.driver_config.num_rows"),
+    pt.get<uint8_t>("aie_metadata.driver_config.shim_row"),
+    pt.get<uint8_t>("aie_metadata.driver_config.reserved_row_start"),
+    pt.get<uint8_t>("aie_metadata.driver_config.reserved_num_rows"),
+    pt.get<uint8_t>("aie_metadata.driver_config.aie_tile_row_start"),
+    pt.get<uint8_t>("aie_metadata.driver_config.aie_tile_num_rows"));
+
+  /* TODO get partition id and uid from XCLBIN or PDI, currently not supported*/
+  uint32_t partition_id = 1;
+  uint32_t uid = 0;
+  drm_zocl_aie_fd aiefd = { partition_id, uid, 0 };
+  int ret = ioctl(mKernelFD, DRM_IOCTL_ZOCL_AIE_FD, &aiefd) ? -errno : 0;
+  if (ret)
+    throw xrt_core::error(ret, "Create AIE failed. Can not get AIE fd");
+
+  ConfigPtr.PartProp.Handle = aiefd.fd;
+
+  AieRC rc;
+  XAie_InstDeclare(DevInst, &ConfigPtr);
+  if ((rc = XAie_CfgInitialize(&DevInst, &ConfigPtr)) != XAIE_OK)
+    throw xrt_core::error(-EINVAL, "Failed to initialize AIE configuration: " + std::to_string(rc));
+  devInst = &DevInst;
+  if(!devInst)
+    throw xrt_core::error(-EINVAL, "Invalid aie object");
+
+  row+=1;
+  auto max_row = pt.get<uint32_t>("aie_metadata.driver_config.num_rows");
+  auto max_col = pt.get<uint32_t>("aie_metadata.driver_config.num_columns");
+  if((row <= 0) && (row >= max_row) && (col < 0) && (col >= max_col))
+    throw xrt_core::error(-EINVAL, "Invalid row or column");
+
+  auto it = regmap.find(v);
```

Comment:
Nice, using a `find()` instead of brackets "[]'.    

Note: If brackets are used, then it will ALWAYS return back a valid entry, even if it is empty.